torch/_functorch/partitioners.py

```diff
@@ -1295,7 +1478,148 @@ def choose_saved_values_set(
+    # The optimized choice is smaller than the inputs anyways
+    if max_act_size <= min_act_size:
+        return runtime_optimized_saved_values
+
+    def get_normalized_size(sz):
+        return (sz / 1e9) / (max_act_size - min_act_size)
+
+    def get_mem_ratio(activations: List[fx.Node]):
+        return (estimate_activations_size(activations) - min_act_size) / (
+            max_act_size - min_act_size
+        )
+
+    more_aggressive_options = replace(
+        min_cut_options,
+        ban_if_used_far_apart=False,
+        ban_if_long_fusible_chains=False,
+        ban_if_materialized_backward=False,
+    )
+    more_aggressive_saved_values, _ = get_saved_values(
+        joint_graph, node_info, more_aggressive_options
+    )
+    if get_mem_ratio(more_aggressive_saved_values) < memory_budget:
+        return more_aggressive_saved_values
+
+    aggressive_options = replace(
+        more_aggressive_options,
+        ban_if_not_in_allowlist=False,
+    )
+    aggressive_recomputation_saved_values, banned_nodes = get_saved_values(
+        joint_graph, node_info, aggressive_options
+    )
+
+    if get_mem_ratio(aggressive_recomputation_saved_values) < memory_budget:
+        return aggressive_recomputation_saved_values
+
+    from torch._inductor.fx_utils import get_node_storage
+
+    input_storages = {get_node_storage(node) for node in node_info.inputs}
+
+    def get_recomputable_banned_nodes(banned_nodes, saved_values):
+        return [
+            i
+            for i in banned_nodes
+            if (i.dist_from_bw < int(1e9) and get_node_storage(i) not in input_storages)
+        ]
+
+    recomputable_banned_nodes = get_recomputable_banned_nodes(
+        banned_nodes, aggressive_recomputation_saved_values
+    )
+
+    # default: runtime_optimized_saved_values
+    # more aggressive: more_aggressive_saved_values
+    # full aggressive: aggressive_recomputation_saved_values
+
+    all_recomputable_banned_nodes = sorted(
+        recomputable_banned_nodes, key=_size_of, reverse=True
+    )
+    if len(all_recomputable_banned_nodes) == 0:
+        return node_info.inputs
+    memories = [get_normalized_size(_size_of(i)) for i in all_recomputable_banned_nodes]
+    runtimes = [estimate_runtime(node) for node in all_recomputable_banned_nodes]
+    cur_memory_budget = memory_budget
+    from torch.utils._mode_utils import no_dispatch
+
+    def get_saved_values_knapsack(memory_budget):
+        with no_dispatch():
+            (
+                expected_runtime,
+                saved_node_idxs,
+                recomputable_node_idxs,
+            ) = _optimize_runtime_with_given_memory(
+                memories, runtimes, max(memory_budget, 0)
+            )
+        dont_ban = set()
+        for idx in recomputable_node_idxs:
+            dont_ban.add(all_recomputable_banned_nodes[idx])
+        predicted_saved_nodes = set(all_recomputable_banned_nodes) - dont_ban
+
+        saved_values, _ = get_saved_values(
```

Comment:
We potentially may call get_saved_values 4 times? Any impact on compilation time?

Reply:
It'll definitely hurt compilation times, but this isn't turned on by default, so I think the effect to compilation times is tolerable for now. We can work on optimizing this in the future.